proofs/signatures/lrat.plf

```diff
@@ -0,0 +1,566 @@
+; The clause had multiple floating literals.
+(declare MRNotUnit MarkResult)
+
+; Determine wether this clause is unit, and if so, modify global assignment to
+; satisfy it.
+;
+; Will never fail.
+;
+; Given a TAUT, where the TAUT variable is the sole floating one, marks that
+; var to satisify its first literal
+(program clause_verify_unit_and_mark ((c clause)) MarkResult
+         (match (clause_is_sat c)
+                (tt MRSat)
+                (ff (match (clause_is_unsat c)
+                           (tt MRUnsat)
+                           (ff ; Dedent
+         (match (clause_has_floating c)
+                (tt (let first (clause_first_floating c)
+                      (do (lit_mk_sat first)
+                        (match (clause_has_floating c)
+                               (tt (do (lit_un_mk_sat first) MRNotUnit))
+                               (ff (MRUnit first))))))
+                ; Unreachable. If clause is not floating it must have been SAT or UNSAT.
+                (ff (fail MarkResult))
+                ))))))
+
+; The return type for the process of Trace-guided unit propegation
+(declare UPResult type)
+; The trace guided unit propegation correctly, but that unit propegation did not end in an empty clause
+(declare UPR_Ok UPResult)
+; The trace guided unit propegation correctly to an empty clause
+(declare UPR_Bottom UPResult)
+; The trace was malformed,
+;; i.e. at some point indicates that a non-unit, non-empty clause should be examined
+(declare UPR_Broken UPResult)
+
+; Execute the unit propegation  indicated by the trace. Report whether that
+; unit propegation succeeds and produces bottom, fails, or succeeds but does
+; not produce bottom.
+; Will never fail.
+; If a TAUT is present, will in some cases allow a TAUT to propegate on of the
+; variables its TAUT on, to SAT that variables first literal.
+(program do_up ((cs CMap) (t Trace)) UPResult
+         (match
+           t
+           (Tracen UPR_Ok)
+           ((Tracec i r) (match (clause_verify_unit_and_mark (CMap_get i cs))
+                                ((MRUnit l)
+                                 (let res (do_up cs r)
+                                   (do (lit_un_mk_sat l) res)))
+                                (MRUnsat UPR_Bottom)
+                                (MRSat UPR_Broken)
+                                (MRNotUnit UPR_Broken)))))
+
+
+; Determine whether a list of indices agrees with the list of indices latent in
+; a list of hints. Both lists should be sorted.
+(program resolution_targets_match (
+                                   (computed CIList)
+                                   (given RATHints)) bool
+         (match given
+                (RATHintsn
+                  (match computed
+                         (CIListn tt)
+                         ((CIListc a b) ff)))
+                ((RATHintsc hint_idx t given')
+                 (match computed
+                        ((CIListc comp_idx computed')
```

Comment:
Good catch. It should return `ff`.

Reply:
Couldn't there be a case (acording to LRAT standard) where `given` is contained but not equal to `computed`? For example, if there are several ways to obtain bottom, shouldn't the hints contain only one way rather than union of all clauses that occur in some proof of bottom?

Reply:
Hmm. If I understand the definition of RAT correctly, for a clause to be a RAT, _all_ of the resolvents have to be AT, not just some of them.
So, if there is a computed resolvent that we do not have a UP trace for, then the proof should not be considered valid.

Reply:
Right! I was confusing hints with traces.